pkg/types/types.go

```diff
@@ -53,18 +55,22 @@ type RetrievalRequest struct {
 	LinkSystem  ipld.LinkSystem
 }
 
-type Retriever func(ctx context.Context, request RetrievalRequest, events func(RetrievalEvent)) (*RetrievalStats, error)
-type CandidateRetriever func(ctx context.Context, request RetrievalRequest, candidates []RetrievalCandidate, events func(RetrievalEvent)) (*RetrievalStats, error)
-type RetrievalCandidateFinder func(ctx context.Context, request RetrievalRequest, events func(RetrievalEvent)) ([]RetrievalCandidate, error)
-
-func WithCandidates(retrievalCandidateFinder RetrievalCandidateFinder, childRetriever CandidateRetriever) Retriever {
-	return func(ctx context.Context, request RetrievalRequest, events func(RetrievalEvent)) (*RetrievalStats, error) {
-		candidates, err := retrievalCandidateFinder(ctx, request, events)
-		if err != nil {
-			return nil, err
-		}
-		return childRetriever(ctx, request, candidates, events)
-	}
+type CandidateStream []RetrievalCandidate
```

Comment:
IMO calling this a stream is confusing. I'd rather see `[]RetrievalCandidate` and know it's just a list of `RetrievalCandidates` than see `CandidatesStream` and think that there it has some kind of streaming functionality.

Reply:
you're right -- I saved the update for the other ticket when it actually is a stream.